assets/model_monitoring/components/src/token_statistics_compute_metrics/compute_token_statistics_metrics.py

```diff
@@ -0,0 +1,351 @@
+    stack_columns = ["'"+ metric_prefix +"_avg_" + col + "', " + metric_prefix + "_avg_" + col for col in columns]
+    stack_expr = "stack(" + str(len(columns)) + ", " + ", ".join(stack_columns) + ") as (metric_name, metric_value)"
+    return token_df.filter(token_df["status_code"] == 200).groupBy(dimensions).agg(*avg_columns).selectExpr("group", "group_pivot", stack_expr)
+
+def compute_sum_df(token_df, columns, dimensions, metric_prefix=""):
+    '''
+    Needs the token_df with columns dimensions, columns
+    Returns a metric df with columns group, group_pivot, metric_name, metric_value
+    where metric_name is the column name and metric_value is the sum of the column
+    '''
+    # check if the token_df has the expected columns
+    if set(dimensions + columns + ['status_code']).issubset(token_df.columns) == False:
+        return
+    # restrict the token_df to only expected columns
+    token_df = token_df.select(dimensions + columns + ['status_code'])
+
+    sum_columns = [sum(col).cast("float").alias(f"{metric_prefix}_sum_{col}") for col in columns]
+    stack_columns = ["'"+ metric_prefix +"_sum_" + col + "', " + metric_prefix + "_sum_" + col for col in columns]
+    stack_expr = "stack(" + str(len(columns)) + ", " + ", ".join(stack_columns) + ") as (metric_name, metric_value)"
+    return token_df.filter(token_df["status_code"] == 200).groupBy(dimensions).agg(*sum_columns).selectExpr("group", "group_pivot", stack_expr)
+
+
+def compute_percentage_metric_df(numerator_metric_df, denominator_metric_df, ratio_metric_name, dimensions):
+    '''
+    Needs the metric df with columns group, group_pivot, metric_name, metric_value. 
+    The numerator_metric_df and denominator_metric_df should have the same number of rows. 
+    Returns a metric df with columns group, group_pivot, metric_name, metric_value,
+    where metric_name is the ratio_metric_name and 
+    metric_value is the ratio of the metric_value of the numerator_metric_df to the metric_value of the denominator_metric_df
+    null metric_values in the numerator_metric_df are replaced with 0
+    '''
+
+    # check if the numerator_metric_df has more number of rows as the denominator_metric_df
+    # this is the error case as we expect a denominator row for each numerator row
+    if numerator_metric_df.count() > denominator_metric_df.count():
+        print(f"Error: The number of rows in the numerator_metric_df is greater the number of rows in the denominator_metric_df")
+        print(f"numerator_metric_df count: {numerator_metric_df.count()}")
+        print(f"numerator_metric_df sample: {numerator_metric_df.take(10)}")
+        print(f"denominator_metric_df count: {denominator_metric_df.count()}")
+        print(f"denominator_metric_df sample: {denominator_metric_df.take(10)}")
+        return spark.createDataFrame([], numerator_metric_df.schema)
+
+    ratio_df = numerator_metric_df.withColumnRenamed("metric_value","numerator")\
+    .join(
```

Comment:
nit: can you indent this to the same depth as the previous line (`numerator_metric_df`) please?